src/include/devcomm.h

```diff
@@ -116,6 +116,14 @@ struct ncclRing {
   int* devUserRanks;
 };
 
+#define SCKL_MAX_NUM_CONN 16
```

Comment:
Is 16 enough? In a cross-node DGX-2 scenario one rank might communicate with all of the local GPUs + a few remote ones (e.g. a separate GPU for each level of a distance-doubling algorithm).

Edit: I see that the connections are currently established between all peers, so I assume this is meant for just in-node stuff currently. In the future we probably need to check at runtime that this limit isn't exceeded.

Reply:
I agree with Olli - should we just bit the bullet and make this work cross node?   What needs to change?

Reply:
the reason for this choice was to follow what NCCL already uses for such structs. for example they have a similar concept for tree and the maximum number of children. Note that this is *per channel*. So, with 2 channels and we have 32 max connections. Of course this is a simple change. Making it a dynamic value requires a lot of changes including malloc and cudaMalloc and freeing them which is inconsistent with NCCL codebase.
@olsaarik, where do you see the connections made with all peers? That comes from an xml example file.

Reply:
Update, I added a guard to not load more connection than it can handle. It throws a warning and ignores when they are too many. Any thoughts on if I should make it an error or leave it as a warning?

Reply:
Ok, I thought the XML parsing code was connecting to all peers (there was a TODO there iirc that sounded like that).

A warning sounds fine. An algorithm later on might error out if the connections it needs aren't there.

I don't think we need to do dynamic allocation for these in any case. Even though something like VHDD might have a dynamic number of connections, it's logarithmic so we can enable a pretty large bound with a relatively small value here.

Reply:
That is correct, once I added another XML parser to parse the algorithm, I will have the necessary check you are pointing out.

Another idea would be to extract the necessary connections out of the algorithm itself and make the entire process in one XML file.